[GH-21] Add support markdown preview in add todo modal

asimsedhain opened the pull request:
<!-- Thank you for contributing a pull request! Here are a few tips to help you:

1. If this is your first contribution, make sure you've read the Contribution Checklist https://developers.mattermost.com/contribute/getting-started/contribution-checklist/
2. Read our blog post about "Submitting Great PRs" https://developers.mattermost.com/blog/2019-01-24-submitting-great-prs
3. Take a look at other repository specific documentation at https://developers.mattermost.com/contribute
-->

#### Summary
<!--
A description of what this pull request does.
-->
This adds support markdown preview in the add todo modal.
![temp1](https://user-images.githubusercontent.com/38435962/105132389-78a8bb80-5ab0-11eb-820d-7e46ffe4542c.png)
![temp2](https://user-images.githubusercontent.com/38435962/105132396-7b0b1580-5ab0-11eb-88b6-fcfa6afb4dbb.png)


#### Ticket Link
<!--
If this pull request addresses a Help Wanted ticket, please link the relevant GitHub issue, e.g.

  Fixes https://github.com/mattermost/mattermost-server/issues/XXXXX

Otherwise, link the JIRA ticket.
-->
Fixes https://github.com/mattermost/mattermost-plugin-todo/issues/21

mattermod commented:
Hello @asimsedhain,

Thanks for your pull request! A Core Committer will review your pull request soon. For code contributions, you can learn more about the review process [here](https://developers.mattermost.com/contribute/getting-started/code-review/).

larkox reviewed (approved):
Wonderful work! Thank you so much for the contribution. LGTM.

mickmister commented on webapp/src/components/root/root.jsx:
```diff
@@ -82,40 +93,90 @@ export default class Root extends React.Component {
+                                    this.setState({message: e.target.value})
+                                }
+                            />
+                        )}{' '}
```
I think this looks better with space on its own line. Also
```suggestion
                        )}
                        {' '}
```

mickmister reviewed (approved):
Some of the multiline formatting changes to the render method's JSX props do not have to do with this PR's functionality. This can mess with the commit tracking of different lines, as it makes it more difficult to quickly access a commit that introduced some functionality.

I also personally think most of the remaining multiline changes they are better to leave single line. It keeps the code more concise to have less lines. The multiline code is straying away from the conventions we usually have in the different javascript projects. Please let me know your thoughts @asimsedhain 🙂

mickmister reviewed (changes requested):
Sorry, didn't mean to hit approve. I have one request regarding creating the class names for the buttons, and a few requests regarding formatting.

Thanks for this great improvement @asimsedhain! 🚀

asimsedhain commented:
Thank you for the feedback, @mickmister.
I forgot I had auto formatted the code. Force of habit.
Also, I have refactored out the class names. Nice catch on that one.
I have incorporated the requested changes. Please let me know your thoughts on it.

mickmister commented on webapp/src/components/root/root.jsx:
```diff
@@ -85,12 +92,37 @@ export default class Root extends React.Component {
+                                onClick={() => {
+                                    this.setState({previewMarkdown: false});
+                                }}
+                            >{'Write'}</button>
```
```suggestion
                            >
                                {'Write'}
                            </button>
```

mickmister commented on webapp/src/components/root/root.jsx:
```diff
@@ -85,12 +92,37 @@ export default class Root extends React.Component {
+                                onClick={() => {
+                                    this.setState({previewMarkdown: true});
+                                }}
+                            >{'Preview'}</button>
```
```suggestion
                            >
                                {'Preview'}
                            </button>
```

mickmister commented on webapp/src/components/root/root.jsx:
```diff
@@ -85,12 +92,37 @@ export default class Root extends React.Component {
+                            <div
+                                className='todoplugin-input'
+                                style={style.markdown}
+                            >{PostUtils.messageHtmlToComponent(
```
```suggestion
                            >
                                {PostUtils.messageHtmlToComponent(
```

mickmister reviewed (approved):
Thanks for addressing my comments @asimsedhain! LGTM, I have a few non-blocking suggestions to improve the formatting of embedded text

aaronrothschild reviewed (approved):
Thanks for your contribution @asimsedhain !  Looks great, looking forward to using this - and making it easier for others to format lists :)

mattermod commented:
This PR has been automatically labelled "stale" because it hasn't had recent activity.
A core team member will check in on the status of the PR to help with questions.
Thank you for your contribution!

/cc @jasonblais @jfrerich @emilyacook

DHaussermann commented:
/update-branch

DHaussermann commented:
@asimsedhain This is a very nice enhancement to the pluign. Well done! It seems to work as expected. I find no functional issues here. 
The only follow-up I have is about the CSS styling. It seems like the colors don't adjust to respect the theme the user has applied. This becomes an issue with dark themes.

![Screen Shot 2021-02-10 at 3 27 08 PM](https://user-images.githubusercontent.com/35588783/107568682-58eb4b80-6bb5-11eb-8e7c-ce159656655b.png)

I'm not sure if this is an existing issue with the plugin that's more apparent now that you're adding functionality or if there is a way the styling can be made more consistent?

cc @larkox  if you have any thoughts on this.

larkox commented:
@DHaussermann Looking at the code, it seems that it is indeed taking the colors from the theme. `Mattermost` and `Windows Dark` themes have almost the same button colors, and that may be why the confusion. The "Add Todo" does indeed have a different logic on the theme and should be updated, but it is still going along with the theme. Have you confirmed that with other themes (like `Mattermost Dark`) the color don't change?

DHaussermann commented:
@larkox yes to clarify, the colours do change when the theme is applied. It's just that ToDo plugin seems to use slightly different colour scheme. I don't think white like on the `Preview` tab above is normally seen on dark theme. Another example is the unselected tab for `Mattermost Dark`
![Screen Shot 2021-02-11 at 9 02 21 AM](https://user-images.githubusercontent.com/35588783/107647513-3b67c180-6c49-11eb-84cc-5deb57487248.png)

So, I was curious if this colour selection is defined explicitly in this PR orif this is more of an issue with the ToDo Css in general.

mattermod commented:
This PR has been automatically labelled "stale" because it hasn't had recent activity.
A core team member will check in on the status of the PR to help with questions.
Thank you for your contribution!

/cc @jasonblais @jfrerich @emilyacook

DHaussermann commented:
Gentle ping on this @larkox 
Is the CSS styling separate form this PR? If so, I'd be happy to create a new issue to address it and we can prceed with this PR now.

larkox commented:
I think we can assume it is separated from this PR indeed. Thanks @DHaussermann !

DHaussermann reviewed (approved):
Tested and passed
Tested various markdown types and the preview is working as expected.
Added TM4J test case to cover this for release testing

Crated follow-up issue here to address some theme colours 
https://github.com/mattermost/mattermost-plugin-todo/issues/140 


Thanks @asimsedhain this is a great enhancement!
